File: worker/ggrstats/duels.py
# worker/ggrstats/duels.py
"""The private races inside the fleet. A duel is two boats next to each other in the ranking, both with a current fix, within
DUEL_NM of each other in distance to finish. For each pair: the gap now, the gap at every 4-hour report of the last three
days (fix to fix, so a missed report is a hole in the line and never a false gain), how often the lead changed hands, when
the boat now ahead got ahead, and how far apart the two are on the water and on which side. Pure functions on plain dicts."""
import bisect, math
from .grid import gc_nm, bearing_deg, SLOT_S

DAY = 86400
DUEL_NM = 15.0
COMPASS = ("N", "NE", "E", "SE", "S", "SW", "W", "NW")

def _dtf_near(fixes, ats, t):
    """Distance to finish (nm) from the latest fix at or before t (+20 min), if that fix is less than 3.5 h old."""
    i = bisect.bisect_right(ats, t + 1200) - 1
    if i < 0 or t - fixes[i]["at"] >= 3.5 * 3600 or not fixes[i].get("dtf"):
        return None
    return fixes[i]["dtf"] / 1852.0

def find(boats, fixes_by_team, T, within_nm=DUEL_NM, days=3):
    fresh = sorted((b for b in boats if not b["stale"]), key=lambda b: b["dtf_nm"])
    out = []
    for a, b in zip(fresh, fresh[1:]):                       # a is ahead of b
        gap = b["dtf_nm"] - a["dtf_nm"]
        if gap > within_nm:
            continue
        fa, fb = fixes_by_team[a["id"]], fixes_by_team[b["id"]]
        ata, atb = [f["at"] for f in fa], [f["at"] for f in fb]
        series = []
        for t in range(T - days * DAY, T + 1, SLOT_S):
            da, db_ = _dtf_near(fa, ata, t), _dtf_near(fb, atb, t)
            if da is not None and db_ is not None:
                series.append([t, db_ - da])                 # positive: today's leader of the pair was ahead then too
        at = {t: g for t, g in series}
        changes, passed_at, prev = 0, None, 0
        for t, g in series:
            s = (g > 0) - (g < 0)
            if s and prev and s != prev:
                changes += 1
                passed_at = t if s > 0 else None             # the moment today's leader of the pair got ahead (for good, so far)
            if s:
                prev = s
        brg = bearing_deg(a["lat"], a["lon"], b["lat"], b["lon"])
        water = gc_nm(a["lat"], a["lon"], b["lat"], b["lon"])
        out.append({"ahead_id": a["id"], "behind_id": b["id"], "gap_nm": gap, "gap24_nm": at.get(T - DAY), "gap72_nm": at.get(T - days * DAY),
                    "lead_changes": changes, "passed_at": passed_at, "water_nm": water,
                    "side": COMPASS[int((brg + 22.5) // 45) % 8] if water >= 0.5 else None, "series": series})
    return out

File: site/lib/duels.ts
// site/lib/duels.ts — one sentence for a duel, from its numbers (worker/ggrstats/duels.py). Pure, tested.
// gap values are "boat behind minus boat ahead" in nm, so a negative past value means today's leader of the pair was behind then.
import { dayMon, hhmm } from "./format";
export type DuelNumbers = { gap_nm: number; gap24_nm: number | null; gap72_nm: number | null; lead_changes: number; passed_at: string | null };
export function duelStory(d: DuelNumbers, ahead: string, behind: string): string {
  const was = d.gap72_nm ?? d.gap24_nm, span = d.gap72_nm != null ? "three days" : "24 hours";
  if (d.passed_at) {
    const when = `${ahead} got ahead at ${hhmm(d.passed_at)} on ${dayMon(d.passed_at)}`;
    if (d.lead_changes >= 2) return `${when}. The lead has changed hands ${d.lead_changes} times in three days.`;
    return d.gap72_nm != null && d.gap72_nm <= -5 ? `${when}, after trailing by ${Math.round(-d.gap72_nm)} nm three days ago.` : `${when}.`;
  }
  if (was == null) return "Too few reports from both boats to tell the story yet.";
  const moved = Math.round(was - d.gap_nm);                       // positive: the boat behind has closed
  if (moved >= 3) return `${behind} has closed ${moved} nm in ${span}.`;
  if (moved <= -3) return `${ahead} has pulled out ${-moved} nm in ${span}.`;
  return `Nothing in it: the gap has moved ${Math.abs(moved)} nm in ${span}.`;
}
